datafusion/physical-expr/src/rewrite.rs

```diff
@@ -113,6 +113,21 @@ pub trait TreeNodeRewritable: Clone {
         Ok(new_node)
     }
 
+    fn mutable_transform_up<F>(self, op: &mut F) -> Result<Self>
+    where
+        F: FnMut(Self) -> Result<Option<Self>>,
+    {
+        let after_op_children =
+            self.map_children(|node| node.mutable_transform_up(op))?;
+
+        let after_op_children_clone = after_op_children.clone();
+        let new_node = match op(after_op_children)? {
+            Some(value) => value,
+            None => after_op_children_clone,
+        };
+        Ok(new_node)
+    }
+
```

Comment:
I think you can leverage the `transform_using()` (Visitor pattern) to achieve the same effects instead of adding another` mutable_transform()` method. 
I think the Visitor pattern is more straightforward and easy to understand if we want to collect some infos and modify the plan/structs during the traversal process.

Reply:
Added, you can check it.